mlir/lib/Dialect/Vector/Transforms/VectorDropLeadUnitDim.cpp

```diff
@@ -346,6 +349,12 @@ mlir::vector::castAwayContractionLeadingOneDim(vector::ContractionOp contractOp,
   // greedily to drop more.
   int64_t dropDim = 1;
 
+  if (isMasked) {
```

Comment:
There are some snippets under the comment `Vector mask setup` or similar that doing something like this. I think we should perhaps create a utility for those and define some kind of canonical form to do this at pattern rewrite level. For conversions, the goal when we introduced the mask op was to try to reduce this kind of conditional code and provide some infra to make it as transparent as possible, hence the mask conversion pattern class that we use for LLVM conversion. Let's brainstorm a bit about the different options and see what can be improved in this regard.

Reply:
@dcaballe I was implementing a general purpose implementation of this, something like https://github.com/tanmaysachan/llvm-project/commit/242f9010fc0b3bf782abcf333b52e1c30121d252

Does this look okay to you? Can refactor for common code between the masked vs non-masked functions to be a function, and the pattern rewrites handle the entry points.

Reply:
Sorry, returning to this after two weeks of being OOO.

@tanmaysachan , from a quick scan, you have adopted [VectorMaskOpConversionBase](https://github.com/llvm/llvm-project/blob/982e9022ca4abaad58c693d2b0aba0e908ee2d39/mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp#L849-L870) which matches `vector.mask`. IIUC, that's going to be insufficient in this case - we need something that would work both for:
*  `vector.mask {vector.contract}`, and 
* `vector.contract`.

`VectorMaskOpConversionBase` would only work for the first case. Perhaps @dcaballe had something else in mind, but IMHO we need another "base" class to accomodate for that. This a blocker for me, so I'm proposing one here (*):
* https://github.com/llvm/llvm-project/pull/83827

> I think we should perhaps create a utility for those and define some kind of canonical form to do this at pattern rewrite level.

Does #83827 make sense? Happy to try something else :) Also, regardless of the long-term approach that we take here, would you be OK with me landing this to unblock `linalg.mmt4d` investigation? 

(*) Apologies @tanmaysachan if you are also actively working on this, but this is quite urgent for me. Mindful of delays due to time difference, I decided to go ahead and draft something quickly.